### DISCUSSION TITLE
Need to use OAuth with redirect

### DISCUSSION THREAD

**Discussion by nazarigue** [fastapi-users/fastapi-users#676] [Q&A]
Hey,

I am stuck a bit.

I want to create google auth and all is working well with your setup. But if i add redirect_url, I don't receive any cookie and not able to retrieve any token.

My flow is:

1. User signs in with Google SSO.
2. Gets redirected to  http://localhost:8000/feed

```
google_oauth_router = fastapi_users.get_oauth_router(google_oauth_client, SECRET,redirect_url='http://localhost:8000/feed')

callback=router.include_router(google_oauth_router, prefix="/auth/google", tags=["auth"])
```

It redirects, but I don't have any info on user.

What to do?

**Comment by frankie567** [comment]
Hi there,

Sorry for the delayed response 🙂

You shouldn't override the redirect URL in your router. We provide a [`/callback` route](https://frankie567.github.io/fastapi-users/usage/routes.html#get-callback) which purpose is to handle the OAuth2 protocol and register or login the user. The effect of this is to create a User in your database with the associated Google access tokens.

It's up to you then to use those tokens to make clever things, like retrieving personal data.

**Reply by nazarigue** [reply]
@frankie567 thanks! I know you have the callback url and works perfectly fine. The response is null on auth/google/callback - how can i retrieve cookie? Or how should I handle the flow when I need to end up not at auth/google/callback  but at www.site.com/feed with credentials returned?

**Reply by frankie567** [reply]
FastAPI Users is designed to be used as a REST API.

Thus, the response doesn't contain anything but, if you are using Cookie authentication, it'll set a proper cookie for the current user. If called in a browser, the browser will then store this cookie and send it in subsequent requests.

To handle this nicely in a frontend website, what you could do is to call the `/callback` endpoint yourself from your frontend app. For example, the redirect URL could point to your `/feed` page. At that point, you can retrieve the `code` in the query parameters and call the `/callback` route yourself using an AJAX request.

Another solution is to customize the `get_login_response` method of the `CookieAuthentication` backend. This method is responsible for setting the proper cookie in the response. You can add some logic to perform a 302 redirect to your `/feed` page. I proposed an example some time ago here: https://github.com/frankie567/fastapi-users/issues/434#issuecomment-754444666

**Reply by davidbrochart** [reply]
> Another solution is to customize the `get_login_response` method of the `CookieAuthentication` backend. This method is responsible for setting the proper cookie in the response. You can add some logic to perform a 302 redirect to your `/feed` page. I proposed an example some time ago here: [#434 (comment)](https://github.com/frankie567/fastapi-users/issues/434#issuecomment-754444666)

I tried this solution but after redirection it looks like the cookie isn't set (`/users/me` gives a `401`).
If I don't customize `get_login_response` and instead enter manually the redirect URL, it works fine.
Any thought?

**Reply by davidbrochart** [reply]
Ah, it was due to the parent `get_login_response()` not being awaited. It works like a charm with the following, thanks!

```python
from fastapi import status
from fastapi_users.authentication import CookieAuthentication

class AutoRedirectCookieAuthentication(CookieAuthentication):
    async def get_login_response(self, user, response):
        await super().get_login_response(user, response)
        response.status_code = status.HTTP_302_FOUND
        response.headers["Location"] = "http://www.myapp.com/route/after/oauth"
```

**Reply by frankie567** [reply]
Oh yes, nice catch, sorry! Glad you made it work 😉

**Reply by nazarigue** [reply]
@frankie567 thanks! In this case, is it possible to request "profile image" from google response or I can have only email?
